src/main/scala/datalog/execution/JoinIndexes.scala

```diff
@@ -79,7 +85,29 @@ object JoinIndexes {
         ("v", variables(v))
       case c: Constant => ("c", c)
     }
-    new JoinIndexes(bodyVars, constants.toMap, projects, deps, rule)
+
+    // The head atom may not be negated.
+    if (rule(0).negated) throw new Exception("Head atom cannot be negated")
+
+    // All variables in the head atom must be limited.
+    val variablesInPositiveAtoms =
```

Comment:
This also seems like a check that should go in the AST transform passes, not here